File: android/app/src/main/java/app/rentivo/app/AppModel.kt
package app.rentivo.app

import androidx.compose.runtime.Stable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.runtime.staticCompositionLocalOf
import app.rentivo.data.AppDependencies
import app.rentivo.data.DemoSettings
import app.rentivo.domain.DemoError
import app.rentivo.domain.MFAChallenge
import app.rentivo.domain.MobileLoginOutcome
import app.rentivo.domain.PasskeyAssertionPayload
import app.rentivo.domain.PixConfiguration
import app.rentivo.domain.UserProfile
import kotlin.coroutines.cancellation.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.launch

/**
 * App-wide session, navigation and notice state. Port of `ios/Rentivo/App/AppModel.swift`.
 *
 * The iOS type is an `@Observable @MainActor` class; the Compose equivalent is a plain state holder
 * whose properties are `mutableStateOf`-backed, created once by `MainActivity` and published to the
 * tree through [LocalAppModel]. Every mutation therefore happens on the main thread, matching the
 * `@MainActor` isolation.
 *
 * @param dependencies the 17 repositories every screen resolves through.
 * @param sessionExpired fires when `LiveAPIClient` sees a 401 for the stored token (the analog of
 *   the iOS `liveAPIClientSessionExpired` notification). Only collected for live dependencies.
 * @param scope the app model's own lifetime, outliving any one screen. It carries the session-expiry
 *   subscription and the session-ending flows ([signOut], [deleteAccount]), which must survive the
 *   screen that started them disappearing.
 */
@Stable
class AppModel(
  val dependencies: AppDependencies,
  sessionExpired: Flow<Unit>? = null,
  private val scope: CoroutineScope,
) {

  sealed interface Session {
    data object Restoring : Session

    data object Anonymous : Session

    data class Authenticated(val profile: UserProfile) : Session
  }

  var session: Session by mutableStateOf(Session.Anonymous)

  var selectedTab: AppTab by mutableStateOf(AppTab.HOME)

  var notice: AppNotice? by mutableStateOf(null)

  var isSigningOut: Boolean by mutableStateOf(false)

  var isDeletingAccount: Boolean by mutableStateOf(false)

  var demoSettings: DemoSettings by mutableStateOf(dependencies.demo.demoSettings)

  var dataRevision: Int by mutableStateOf(0)

  init {
    if (dependencies.auth.usesLiveAPI) {
      session = Session.Restoring
      observeSessionExpiry(sessionExpired)
    }
  }

  val currentUser: UserProfile
    get() = (session as? Session.Authenticated)?.profile ?: dependencies.auth.currentUser

  val isAuthenticated: Boolean
    get() = session is Session.Authenticated

  val usesLiveAPI: Boolean
    get() = dependencies.auth.usesLiveAPI

  suspend fun loadProfile(): UserProfile {
    val profile = dependencies.profile.profile()
    if (session is Session.Authenticated) {
      session = Session.Authenticated(profile)
    }
    return profile
  }

  suspend fun updateProfilePIX(pix: PixConfiguration): UserProfile {
    val profile = dependencies.profile.updatePix(pix)
    if (session is Session.Authenticated) {
      session = Session.Authenticated(profile)
    }
    return profile
  }

  suspend fun restoreSessionIfNeeded() {
    // Only a live-API session ever starts out `Restoring` (see `init`), so the demo store returns
    // here before its `restoreSession()` — which has nothing to restore — runs.
    if (session !is Session.Restoring) return
    session = try {
      dependencies.auth.restoreSession()?.let(Session::Authenticated) ?: Session.Anonymous
    } catch (cancellation: CancellationException) {
      throw cancellation
    } catch (throwable: Throwable) {
      notice = AppNotice(
        kind = AppNotice.Kind.WARNING,
        message = "Não foi possível restaurar sua sessão. Entre novamente.",
      )
      Session.Anonymous
    }
  }

  fun signIn() {
    session = Session.Authenticated(currentUser)
    selectedTab = AppTab.HOME
    notice = AppNotice(
      kind = AppNotice.Kind.SUCCESS,
      message = "Bem-vinda à demonstração do Rentivo.",
    )
  }

  // MARK: - Native sign-in
  //
  // These are the app-state half of the native (`/api/v1/auth/mobile/*`) flow the login screen
  // drives. `dependencies.auth` already owns the credential half — every one of these calls
  // persists the bearer token and records the profile as the store's current user before returning
  // — so nothing here re-adopts it; they only move the session, the selected tab, and the notice.
  // Errors propagate to the caller, which owns the screen the user is still looking at; the session
  // is left untouched so a failed attempt keeps them on the form.
  //
  // `signIn` is the only one that can return without a session: its `MfaRequired` outcome is handed
  // back verbatim so the login screen can present the second factor and finish with one of the
  // completion calls below.

  suspend fun signIn(email: String, password: String): MobileLoginOutcome {
    val outcome = dependencies.auth.mobileLogin(email = email, password = password)
    if (outcome is MobileLoginOutcome.Authenticated) adoptSignedInProfile(outcome.profile)
    return outcome
  }

  suspend fun signUp(email: String, password: String) {
    adoptSignedInProfile(dependencies.auth.mobileSignup(email = email, password = password))
  }

  suspend fun completeTOTP(challenge: MFAChallenge, code: String) {
    adoptSignedInProfile(dependencies.auth.verifyTotp(challenge = challenge, code = code))
  }

  suspend fun completeRecoveryCode(challenge: MFAChallenge, code: String) {
    adoptSignedInProfile(dependencies.auth.verifyRecoveryCode(challenge = challenge, code = code))
  }

  suspend fun completePasskey(challenge: MFAChallenge, credential: PasskeyAssertionPayload) {
    adoptSignedInProfile(
      dependencies.auth.completePasskeyAssertion(challenge = challenge, credential = credential)
    )
  }

  /** The state every server-backed sign-in lands on, whichever path produced the profile. */
  private fun adoptSignedInProfile(profile: UserProfile) {
    session = Session.Authenticated(profile)
    selectedTab = AppTab.HOME
    notice = AppNotice(kind = AppNotice.Kind.SUCCESS, message = "Sessão conectada ao Rentivo.")
  }

  /**
   * Signs out, ending on the anonymous screen.
   *
   * Deliberately not a `suspend` function. Its very first effect — [completeSignOut] — removes the
   * account screen that started it from composition, which cancels that screen's
   * `rememberCoroutineScope()`; the token revocation still in flight at that point would be
   * cancelled somewhere between "always runs" and "never runs" depending on frame timing. Running
   * the sequence on the app model's own scope, which lives as long as the session does, makes it
   * complete regardless of what happens to the caller. Callers keep working unchanged: invoking a
   * non-suspending function inside `scope.launch { … }` is fine.
   */
  fun signOut() {
    if (isSigningOut) return
    // Demo mode has no token to revoke, so it drops straight to local state — synchronously, with
    // no scope involved and no in-flight flag to toggle.
    if (!dependencies.auth.usesLiveAPI) {
      completeSignOut()
      return
    }
    // Claimed here rather than inside the coroutine so that a second tap arriving in the same frame
    // is rejected by the guard above instead of starting a competing sign-out.
    isSigningOut = true
    scope.launch {
      try {
        // Revoke the API token first (best-effort inside `logout()`), then unconditionally drop
        // local credentials/state: the user must never end up "signed out" locally while the server
        // still honors the old token, nor stuck signed in locally because a later step failed. With
        // native sign-in there is no browser session to close afterwards.
        dependencies.auth.logout()
        completeSignOut()
      } finally {
        isSigningOut = false
      }
    }
  }

  /** Deletes the account and signs out. Runs on the app model's scope, for the reasons in [signOut]. */
  fun deleteAccount(password: String) {
    if (isDeletingAccount) return
    // There is no demo account to delete, so demo mode just returns to the signed-out screen
    // without claiming a deletion happened.
    if (!dependencies.auth.usesLiveAPI) {
      completeSignOut()
      return
    }
    isDeletingAccount = true
    scope.launch {
      try {
        dependencies.auth.deleteAccount(password = password)
        completeSignOut()
        notice = AppNotice(kind = AppNotice.Kind.SUCCESS, message = "Sua conta foi excluída.")
      } catch (cancellation: CancellationException) {
        throw cancellation
      } catch (throwable: Throwable) {
        notice = AppNotice(
          kind = AppNotice.Kind.WARNING,
          message = DemoError.from(throwable).message,
        )
      } finally {
        isDeletingAccount = false
      }
    }
  }

  private fun completeSignOut() {
    session = Session.Anonymous
    selectedTab = AppTab.HOME
    notice = null
  }

  /**
   * Reacts to `LiveAPIClient` reporting that the stored token is no longer valid (a 401 during a
   * background/foreground request), which otherwise leaves the app "authenticated" while every
   * screen keeps failing until relaunch. This subscription lives for the lifetime of the app model.
   */
  private fun observeSessionExpiry(sessionExpired: Flow<Unit>?) {
    val expiries = sessionExpired ?: return
    scope.launch { expiries.collect { handleSessionExpired() } }
  }

  private fun handleSessionExpired() {
    // A deliberate `signOut()` also revokes the token via the live store's `logout()`, whose POST
    // can itself 401 (the token it's revoking is, after all, about to become invalid) and emit this
    // same event concurrently. Without this guard that race would flash "Sua sessão expirou"
    // mid-sign-out even though the user chose to sign out, not because the session actually expired
    // out from under them.
    if (isSigningOut) return
    if (session !is Session.Authenticated) return
    session = Session.Anonymous
    selectedTab = AppTab.HOME
    notice = AppNotice(
      kind = AppNotice.Kind.WARNING,
      message = "Sua sessão expirou. Entre novamente para continuar.",
    )
  }

  fun showNotice(message: String, kind: AppNotice.Kind = AppNotice.Kind.SUCCESS) {
    notice = AppNotice(kind = kind, message = message)
  }

  fun setDelayEnabled(enabled: Boolean) {
    dependencies.demo.setDelayEnabled(enabled)
    refreshDemoState(reloadContent = false)
  }

  fun setEmptyMode(enabled: Boolean) {
    dependencies.demo.setEmptyMode(enabled)
    refreshDemoState(reloadContent = true)
  }

  fun setViewerMode(enabled: Boolean) {
    dependencies.demo.setViewerMode(enabled)
    refreshDemoState(reloadContent = true)
  }

  fun failNextOperation() {
    dependencies.demo.failNextOperation()
  }

  fun resetDemo() {
    dependencies.demo.reset()
    refreshDemoState(reloadContent = true)
  }

  private fun refreshDemoState(reloadContent: Boolean) {
    demoSettings = dependencies.demo.demoSettings
    if (reloadContent) dataRevision += 1
  }
}

/**
 * The one [AppModel] for the whole tree, provided by `MainActivity`.
 *
 * Screens read it with `val app = LocalAppModel.current` — the Compose analog of the iOS
 * `@Environment(AppModel.self)` — and reload themselves on demo-data changes with
 * `LaunchedEffect(app.dataRevision) { load() }`.
 */
val LocalAppModel = staticCompositionLocalOf<AppModel> {
  error("LocalAppModel was read outside of the app shell; provide it from MainActivity.")
}

File: ios/Rentivo/App/AppModel.swift
import Foundation
import Observation
import SwiftUI

enum AppTab: Hashable {
  case home
  case billings
  case organizations
  case account
}

enum NoticeArea: Hashable, Sendable {
  case authentication
  case home
  case billings
  case billingDetail
  case billOperations
  case organizations
  case invitations
  case account
  case security
  case apiKeys
  case appearance
  case demoScenarios
}

extension AppTab {
  var noticeArea: NoticeArea {
    switch self {
    case .home: .home
    case .billings: .billings
    case .organizations: .organizations
    case .account: .account
    }
  }
}

struct AppNotice: Identifiable, Equatable {
  enum Kind {
    case success
    case information
    case warning
  }

  let id: UUID
  let kind: Kind
  let message: String
  let owner: NoticeArea

  init(
    id: UUID = UUID(),
    kind: Kind,
    message: String,
    owner: NoticeArea = .authentication
  ) {
    self.id = id
    self.kind = kind
    self.message = message
    self.owner = owner
  }
}

struct AppNoticeClock {
  let now: @MainActor () -> TimeInterval
  let sleep: @MainActor (TimeInterval) async throws -> Void

  static let continuous = AppNoticeClock(
    now: { ProcessInfo.processInfo.systemUptime },
    sleep: { duration in
      try await ContinuousClock().sleep(for: .seconds(duration))
    }
  )
}

@MainActor
@Observable
final class AppModel {
  enum Session {
    case restoring
    case anonymous
    case authenticated(UserProfile)
  }

  var session: Session = .anonymous
  var selectedTab: AppTab = .home {
    didSet { activateNoticeArea(selectedTab.noticeArea) }
  }
  private(set) var notice: AppNotice?
  var isSigningOut = false
  var isDeletingAccount = false
  var demoSettings: DemoSettings
  var dataRevision = 0
  let dependencies: AppDependencies
  private(set) var activeNoticeArea: NoticeArea = .authentication
  private let noticeClock: AppNoticeClock
  private var noticeDismissalTask: Task<Void, Never>?
  private var mountedNoticeID: UUID?
  private var noticeRemainingLifetime: TimeInterval?
  private var noticeTimerStartedAt: TimeInterval?
  private var noticeReduceMotion = false

  init(
    store: MockRentivoStore = MockRentivoStore(fixtures: .canonical),
    noticeClock: AppNoticeClock = .continuous
  ) {
    dependencies = .mock(store: store)
    demoSettings = store.demoSettings
    self.noticeClock = noticeClock
  }

  init(dependencies: AppDependencies, noticeClock: AppNoticeClock = .continuous) {
    self.dependencies = dependencies
    self.noticeClock = noticeClock
    demoSettings = dependencies.demo.demoSettings
    if dependencies.auth.usesLiveAPI {
      session = .restoring
      observeSessionExpiry()
    }
  }

  var currentUser: UserProfile {
    if case .authenticated(let user) = session { return user }
    return dependencies.auth.currentUser
  }

  var isAuthenticated: Bool {
    if case .authenticated = session { return true }
    return false
  }

  func loadProfile() async throws -> UserProfile {
    let profile = try await dependencies.profile.profile()
    if case .authenticated = session {
      session = .authenticated(profile)
    }
    return profile
  }

  func updateProfilePIX(_ pix: PixConfiguration?) async throws -> UserProfile {
    let profile = try await dependencies.profile.updatePix(pix)
    if case .authenticated = session {
      session = .authenticated(profile)
    }
    return profile
  }

  func restoreSessionIfNeeded() async {
    // Only a live-API session ever starts out `.restoring` (see `init(dependencies:)`), so the
    // demo store returns here before its `restoreSession()` — which has nothing to restore — runs.
    guard case .restoring = session else { return }
    do {
      session = try await dependencies.auth.restoreSession().map(Session.authenticated) ?? .anonymous
    } catch {
      session = .anonymous
      showNotice(
        "Não foi possível restaurar sua sessão. Entre novamente.",
        kind: .warning,
        owner: .authentication
      )
    }
  }

  var usesLiveAPI: Bool { dependencies.auth.usesLiveAPI }

  func signIn() {
    session = .authenticated(currentUser)
    selectedTab = .home
    showNotice("Bem-vinda à demonstração do Rentivo.", owner: .home)
  }

  // MARK: - Native sign-in
  //
  // These are the app-state half of the native (`/api/v1/auth/mobile/*`) flow the login screen
  // drives.
  //
  // `dependencies.auth` already owns the credential half — every one of these calls persists the
  // bearer token and records the profile as the store's current user before returning (see
  // `MobileAuthRepositoryTests`) — so nothing here re-adopts it; they only move the session,
  // the selected tab, and the notice. Errors propagate to the caller, which owns the screen the
  // user is still looking at; the session is left untouched so a failed attempt keeps them on
  // the form.
  //
  // `mobileLogin` is the only one that can return without a session: its `.mfaRequired` outcome
  // is handed back verbatim so the login screen can present the second factor and finish with
  // one of the completion calls below.

  func signIn(email: String, password: String) async throws -> MobileLoginOutcome {
    let outcome = try await dependencies.auth.mobileLogin(email: email, password: password)
    if case .authenticated(let profile) = outcome { adoptSignedInProfile(profile) }
    return outcome
  }

  func signUp(email: String, password: String) async throws {
    adoptSignedInProfile(try await dependencies.auth.mobileSignup(email: email, password: password))
  }

  func completeTOTP(challenge: MFAChallenge, code: String) async throws {
    adoptSignedInProfile(try await dependencies.auth.verifyTotp(challenge: challenge, code: code))
  }

  func completeRecoveryCode(challenge: MFAChallenge, code: String) async throws {
    adoptSignedInProfile(
      try await dependencies.auth.verifyRecoveryCode(challenge: challenge, code: code))
  }

  func completePasskey(challenge: MFAChallenge, credential: PasskeyAssertionPayload) async throws {
    adoptSignedInProfile(
      try await dependencies.auth.completePasskeyAssertion(
        challenge: challenge, credential: credential))
  }

  /// The state every server-backed sign-in lands on, whichever path produced the profile.
  private func adoptSignedInProfile(_ profile: UserProfile) {
    session = .authenticated(profile)
    selectedTab = .home
    showNotice("Sessão conectada ao Rentivo.", owner: .home)
  }

  func signOut() async {
    guard !isSigningOut else { return }
    // Demo mode has no token to revoke, so it drops straight to local state.
    guard dependencies.auth.usesLiveAPI else {
      completeSignOut()
      return
    }
    isSigningOut = true
    defer { isSigningOut = false }
    // Revoke the API token first (best-effort inside `logout()`),
    // then unconditionally drop local credentials/state: the user must never
    // end up "signed out" locally while the server still honors the old
    // token, nor stuck signed in locally because a later step failed.
    await dependencies.auth.logout()
    completeSignOut()
  }

  func deleteAccount(password: String) async {
    guard !isDeletingAccount else { return }
    // There is no demo account to delete, so demo mode just returns to the signed-out screen
    // without claiming a deletion happened.
    guard dependencies.auth.usesLiveAPI else {
      completeSignOut()
      return
    }
    isDeletingAccount = true
    defer { isDeletingAccount = false }
    do {
      try await dependencies.auth.deleteAccount(password: password)
      completeSignOut()
      // After `completeSignOut()`, which clears the notice the deletion needs to report.
      showNotice("Sua conta foi excluída.", owner: .authentication)
    } catch {
      showNotice(error.localizedDescription, kind: .warning)
    }
  }

  private func completeSignOut() {
    session = .anonymous
    selectedTab = .home
    activeNoticeArea = .authentication
    dismissNotice()
  }

  /// Reacts to `LiveAPIClient` reporting that the stored token is no longer
  /// valid (a 401 during a background/foreground request), which otherwise
  /// leaves the app "authenticated" while every screen keeps failing until
  /// relaunch. This subscription lives for the lifetime of the app model.
  private func observeSessionExpiry() {
    Task { [weak self] in
      for await _ in NotificationCenter.default.notifications(named: .liveAPIClientSessionExpired) {
        self?.handleSessionExpired()
      }
    }
  }

  private func handleSessionExpired() {
    // A deliberate `signOut()` also revokes the token via `liveStore.logout()`, whose POST can
    // itself 401 (the token it's revoking is, after all, about to become invalid) and fire this
    // same notification concurrently. Without this guard that race would flash "Sua sessão
    // expirou" mid-sign-out even though the user chose to sign out, not because the session
    // actually expired out from under them.
    guard !isSigningOut else { return }
    guard case .authenticated = session else { return }
    session = .anonymous
    selectedTab = .home
    showNotice(
      "Sua sessão expirou. Entre novamente para continuar.",
      kind: .warning,
      owner: .authentication
    )
  }

  func showNotice(
    _ message: String,
    kind: AppNotice.Kind = .success,
    owner: NoticeArea? = nil
  ) {
    cancelNoticeTimer()
    mountedNoticeID = nil
    noticeRemainingLifetime = nil
    let animation: Animation = notice == nil
      ? .easeOut(duration: noticeReduceMotion ? 0.15 : 0.22)
      : .easeInOut(duration: 0.15)
    withAnimation(animation) {
      notice = AppNotice(kind: kind, message: message, owner: owner ?? activeNoticeArea)
    }
  }

  func noticeDidMount(id: UUID, voiceOverEnabled: Bool, reduceMotion: Bool = false) {
    guard notice?.id == id, mountedNoticeID != id else { return }
    mountedNoticeID = id
    noticeReduceMotion = reduceMotion
    startNoticeTimer(id: id, duration: voiceOverEnabled ? 8 : 4)
  }

  func dismissNotice(id: UUID? = nil) {
    guard id == nil || notice?.id == id else { return }
    cancelNoticeTimer()
    mountedNoticeID = nil
    noticeRemainingLifetime = nil
    withAnimation(.easeIn(duration: noticeReduceMotion ? 0.15 : 0.16)) {
      notice = nil
    }
  }

  func noticeInteractionBegan(id: UUID) {
    guard notice?.id == id else { return }
    if let startedAt = noticeTimerStartedAt, let remaining = noticeRemainingLifetime {
      noticeRemainingLifetime = max(0, remaining - (noticeClock.now() - startedAt))
    }
    cancelNoticeTimer()
  }

  func noticeInteractionEnded(id: UUID, committed: Bool) {
    guard notice?.id == id else { return }
    if committed {
      dismissNotice(id: id)
    } else {
      startNoticeTimer(id: id, duration: max(noticeRemainingLifetime ?? 1, 1))
    }
  }

  func activateNoticeArea(_ area: NoticeArea) {
    activeNoticeArea = area
    if let notice, notice.owner != area {
      dismissNotice(id: notice.id)
    }
  }

  func sceneDidBecomeInactive() {
    dismissNotice()
  }

  func updateNoticeReduceMotion(_ enabled: Bool) {
    noticeReduceMotion = enabled
  }

  private func startNoticeTimer(id: UUID, duration: TimeInterval) {
    cancelNoticeTimer()
    noticeRemainingLifetime = duration
    noticeTimerStartedAt = noticeClock.now()
    let clock = noticeClock
    noticeDismissalTask = Task { [weak self] in
      do {
        try await clock.sleep(duration)
        try Task.checkCancellation()
        self?.dismissNotice(id: id)
      } catch {
        // Cancellation is the expected result for replacement, manual dismissal, navigation,
        // scene changes, and sign-out. The identity check above protects against custom test
        // clocks that resume an already-cancelled sleep.
      }
    }
  }

  private func cancelNoticeTimer() {
    noticeDismissalTask?.cancel()
    noticeDismissalTask = nil
    noticeTimerStartedAt = nil
  }

  func setDelayEnabled(_ enabled: Bool) {
    dependencies.demo.setDelayEnabled(enabled)
    refreshDemoState(reloadContent: false)
  }

  func setEmptyMode(_ enabled: Bool) {
    dependencies.demo.setEmptyMode(enabled)
    refreshDemoState(reloadContent: true)
  }

  func setViewerMode(_ enabled: Bool) {
    dependencies.demo.setViewerMode(enabled)
    refreshDemoState(reloadContent: true)
  }

  func failNextOperation() {
    dependencies.demo.failNextOperation()
  }

  func resetDemo() {
    dependencies.demo.reset()
    refreshDemoState(reloadContent: true)
  }

  private func refreshDemoState(reloadContent: Bool) {
    demoSettings = dependencies.demo.demoSettings
    if reloadContent { dataRevision += 1 }
  }
}
